pkg/reconciler/taskrun/taskrun_test.go

```diff
@@ -1707,7 +1707,335 @@ spec:
+  namespace: foo
+spec:
+  retries: 1
+  taskRef:
+    name: test-results-task
+status:
+  taskResults:
+  - name: aResult
+    type: string
+    value: aResultValue
+  conditions:
+  - reason: ToBeRetried
+    status: Unknown
+    type: Succeeded
+    message: "missmatched Types for these results, map[aResult:[array]]"
+  sideCars:
+  retriesStatus:
+  - conditions:
+    - reason: TaskRunValidationFailed
+      status: "False"
+      type: "Succeeded"
+      message: "missmatched Types for these results, map[aResult:[array]]"
+    startTime: "2021-12-31T23:59:59Z"
+    completionTime: "2022-01-01T00:00:00Z"
+    podName: "test-taskrun-results-type-mismatched-pod"
+    taskResults:
+    - name: aResult
+      type: string
+      value: aResultValue
+`)
+		reconciliatonError = fmt.Errorf("1 error occurred:\n\t* missmatched Types for these results, map[aResult:[array]]")
+	)
+
+	for _, tc := range []struct {
+		name                    string
+		task                    *v1beta1.Task
+		tr                      *v1beta1.TaskRun
+		expectedTr              *v1beta1.TaskRun
+		pod                     *corev1.Pod
+		reconcileError          error
+		expectCompletionTimeSet bool
+	}{{
+		name:                    "No Retry on Cancellation",
+		task:                    simpleTask,
+		tr:                      toBeCanceledTaskRun,
+		expectedTr:              canceledTaskRun,
+		expectCompletionTimeSet: true,
+	}, {
+		name:       "Retry on TimedOut",
+		task:       simpleTask,
+		tr:         toBeTimedOutTaskRun,
+		expectedTr: timedOutTaskRun,
+	}, {
+		name:       "Retry on TaskRun Pod Failure",
+		task:       simpleTask,
+		tr:         toFailOnPodFailureTaskRun,
+		expectedTr: failedOnPodFailureTaskRun,
+		pod:        failedPod,
+	}, {
+		name:           "Retry on TaskRun Prepare Failure",
+		tr:             toFailOnPrepareTaskRun,
+		expectedTr:     failedOnPrepareTaskRun,
+		reconcileError: prepareError,
+	}, {
+		name:           "Retry on TaskRun Reconciliation Failure",
+		task:           resultsTask,
+		tr:             toFailOnReconcileFailureTaskRun,
+		expectedTr:     failedOnReconcileFailureTaskRun,
+		reconcileError: reconciliatonError,
+	}} {
+		t.Run(tc.name, func(t *testing.T) {
+			d := test.Data{
+				TaskRuns: []*v1beta1.TaskRun{tc.tr},
+				ConfigMaps: []*corev1.ConfigMap{{
+					ObjectMeta: metav1.ObjectMeta{Namespace: system.Namespace(), Name: config.GetFeatureFlagsConfigName()},
+					Data: map[string]string{
+						"enable-api-fields": config.AlphaAPIFields,
+					},
+				}},
+			}
+			if tc.task != nil {
+				d.Tasks = []*v1beta1.Task{tc.task}
+			}
+			if tc.pod != nil {
+				d.Pods = []*corev1.Pod{tc.pod}
+			}
+			testAssets, cancel := getTaskRunController(t, d)
+			defer cancel()
+			c := testAssets.Controller
+			clients := testAssets.Clients
+			createServiceAccount(t, testAssets, "default", tc.tr.Namespace)
+
+			err := c.Reconciler.Reconcile(testAssets.Ctx, getRunName(tc.tr))
+			if tc.reconcileError != nil {
```

Comment:
This code block can be simplified as:

```
// check for requeueError here and set `ok`
if !ok || tc.reconcileError != err {
    t.Errorf("got: %v; want %v", err, tc.reconcileError)
}
```

Reply:
Also: comparing error strings (like 2008 below) is fragile; suggest checking `errors.Is()` instead (which may require you to define an error type as in https://pkg.go.dev/errors

Reply:
Updated! Yeah It's not easy for this PR to use errors.Is() to check the error. We have so many different places to compare error strings which is not very ideal, I think this is a proper future todos to improve our project Health. Will create an issue to capture this.

Reply:
The reason that I'm comparing strings instead of using `errors.Is()` in this PR:

I tried to define var expectedError := errors.New("xx") so that I can use Error.Is() to compare the errors, then I figured out that `Reconcile()` can return different error types (`controller.PermanentError`, `multierror.Error`):
https://github.com/tektoncd/pipeline/blob/f732985ec30817703e6eeaa6914cc77a2b8adf3c/pkg/reconciler/taskrun/taskrun.go#L314-L318

I defined the two types of errors in this test case, one works fine (controller.PermanentError).

However, the "errors.Is()" does not work for `multiperror.Error`", then I used `cmp.Diff()` to see what's the difference, The result indicates that it compares the pointer values:

**The code**

```
wantEr := (&multierror.Error{Errors: []error{errors.New("missmatched Types for these results, map[aResult:[array]]")}}).ErrorOrNil()
err := c.Reconciler.Reconcile(testAssets.Ctx, getRunName(tc.tr))

if d := cmp.Diff(err, wantEr, cmpopts.EquateErrors()); d != "" {
  t.Errorf("got: %v, want: %v, diff: %v", err, tc.wantReconcileError, d)
}
```

**The comparison result**
```
, diff   (*multierror.Error)(
- 	&⟪0xc000525420⟫{
- 		Errors: []error(⟪ptr:0xc000875d10, len:1, cap:1⟫{
- 			&⟪0xc000875bf0⟫"errors".errorString{s: "missmatched Types for these results, map[aResult:[array]]"},
- 		}),
- 	},
+ 	&⟪0xc00088d200⟫{
+ 		Errors: []error(⟪ptr:0xc0003d6480, len:1, cap:1⟫{
+ 			&⟪0xc0003d6440⟫"errors".errorString{s: "missmatched Types for these results, map[aResult:[array]]"},
+ 		}),
+ 	},
  )
```

Okay okay, then, maybe I should retrieve the Error embedded in multierror.Error() out and the use errors.Is(), but to do that, I have to do `type assertion` on that error so that I can use `(*multierror.Error).Unwrap()`. But wait, this is against `DAMP` instead of `DRY` in unit tests: https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

So I just gave up XD, but pls let me know how I can fix this. Otherwise I'd prefer sticking to the string comparison (which is not ideal!) here.